src/backend/executor/execUtils.c

```diff
@@ -2098,6 +2099,21 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
 		ExecSquelchNode(queryDesc->planstate);
 	}
 
+	/*
+	 * If the reading part of SharedScan is in slice 0, then we will not notify
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
```

Comment:
Really hard to understand, I can hardly define what the original problem was about. Rewrite it entirely, please.